### PR TITLE
Feat: JBDirectory controller whitelist

### PR DESCRIPTION
**Pull request by drgorillamd**
image

Only non covered is actually covered :
image

### CODE REVIEW

**Comment by drgorillamd** [comment]
REVIEW/DISCUSS access for `setControllerOf(...)` :
Now, any address can call `setControllerOf` and, as long as `IJBController _controller` is in knownControllers, `requirePermissionAllowingOverride` will be overridden (without, for instance, the funding cycle check and the premigration process our canonical controller is doing)

Should :
```
    requirePermissionAllowingOverride(
      projects.ownerOf(_projectId),
      _projectId,
      JBOperations.SET_CONTROLLER,
      _knownControllers[_controller]
    )
```
    
Rather be : 
```
    requirePermissionAllowingOverride(
      projects.ownerOf(_projectId),
      _projectId,
      JBOperations.SET_CONTROLLER,
      _knownControllers[**msg.sender**]
    )
```
(the knownController calling it being the "new" or current one of a project, so the premigration&migration are done as intended)

File: contracts/JBDirectory.sol [hunk 1]
```diff
@@ -182,27 +182,21 @@ contract JBDirectory is IJBDirectory, JBOperatable, Ownable {
     @param _projectId The ID of the project to set a new controller for.
     @param _controller The new controller to set.
   */
-  function setControllerOf(uint256 _projectId, IJBController _controller)
-    external
-    override
-    requirePermissionAllowingOverride(
```
Comment (drgorillamd) [inline contracts/JBDirectory.sol:178]: extra emphasis on this change:)

File: .gitignore [hunk 2]
```diff
@@ -13,4 +13,6 @@ deployments/localhost/**
 coverage/
 coverage.json
 
-*.txt
\ No newline at end of file
+*.txt
+.env
```
Comment (odd-amphora) [inline .gitignore:17]: .env already exists at the top.

Maybe separate this by

#VSCode
workspace.code-workspace

File: contracts/JBDirectory.sol [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
 import './interfaces/IJBDirectory.sol';
 import './abstract/JBOperatable.sol';
 import './libraries/JBOperations.sol';
 
 /**
   @notice
-  Keeps a reference of which terminal contracts each project is currently accepting funds through, and which controller contract is managing each project's tokens and funding cycles.
+  Keeps a reference of which terminal contracts each project is currently
+  accepting funds through, and which controller contract is managing each
+  project's tokens and funding cycles.
```
Comment (odd-amphora) [inline contracts/JBDirectory.sol:15]: Why this change? Do we have a set column width? (we should)

File: contracts/JBDirectory.sol [hunk 4]
```diff
@@ -32,6 +36,14 @@ contract JBDirectory is IJBDirectory, JBOperatable {
   */
   mapping(uint256 => mapping(address => IJBTerminal)) private _primaryTerminalOf;
 
+  /**
```
Comment (odd-amphora) [inline contracts/JBDirectory.sol:39]: Per my comment above, the line wrap/column width seems inconsistent.

File: contracts/JBDirectory.sol [hunk 5]
```diff
@@ -32,6 +36,14 @@ contract JBDirectory is IJBDirectory, JBOperatable {
...
+    @notice
+    Addresses that can change a project's controllers. Those are controllers
+    that have been vetted and verified by Juicebox owners as well as custom
+    launching project contracts
```
Comment (odd-amphora) [inline contracts/JBDirectory.sol:43]: Consider: Addresses that can set a project's controller. These addresses/contracts have been vetted and verified by Juicebox owners.

File: contracts/JBDirectory.sol [hunk 6]
```diff
@@ -154,43 +178,31 @@ contract JBDirectory is IJBDirectory, JBOperatable {
...
     Update the controller that manages how terminals interact with the ecosystem.
-
     @dev 
     A controller can be set if:
-    - the message sender is the project owner or an operator is changing the controller.
-    - or, the controller hasn't been set yet and the message sender is the controller being set.
-    - or, the current controller is setting a new controller.
-
+    - the message sender is the project owner or an operator having the correct authorization.
+    - or an allowed address is setting a new controller.
```
Comment (odd-amphora) [inline contracts/JBDirectory.sol:187]: nit) or,

File: contracts/JBDirectory.sol [hunk 7]
```diff
@@ -319,4 +331,42 @@ contract JBDirectory is IJBDirectory, JBOperatable {
...
+
+  /** 
+    @notice
+    The owner (Juicebox multisig) can add addresses which are allowed to change
+    a project's controller. Those addresses are known and vetted controllers as well as
+    contracts designed to launch new projects. This is not a requirement for all controllers.
+    However, unknown controllers may require additional transactions to perform certain operations.
+
+    @dev
+    If you would like your JBController allowlisted, please reach out to the Juicebox dev team.
```
Comment (odd-amphora) [inline contracts/JBDirectory.sol:343]: If you would like an address/contract allowlisted

^ remember these are no longer necessarily controllers

File: test/jb_directory/add_known_controller.test.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      jbDirectory.connect(deployer).addToSetControllerAllowlist(mockJbController.address)
+    ).to.emit(jbDirectory, 'AddToSetControllerAllowlist')
+      .withArgs(mockJbController.address, deployer.address);
+
+    expect(
+      await jbDirectory.isAllowedToSetController(mockJbController.address)
+    ).to.be.true;
+  });
+
+  it('Can\'t add known controller if caller is not JBDirectory owner', async function () {
```
Comment (odd-amphora) [inline test/jb_directory/add_known_controller.test.js:46]: If \ characters are in the name, you will need to wrap with ` ` quotes - otherwise the test name will show as, "Can"

File: contracts/JBDirectory.sol [hunk 9]
```diff
@@ -154,43 +178,31 @@ contract JBDirectory is IJBDirectory, JBOperatable {
...
   function setControllerOf(uint256 _projectId, IJBController _controller)
     external
     override
     requirePermissionAllowingOverride(
       projects.ownerOf(_projectId),
       _projectId,
       JBOperations.SET_CONTROLLER,
-      (address(controllerOf[_projectId]) == address(0) && msg.sender == address(_controller)) ||
-        address(controllerOf[_projectId]) == msg.sender
+      (_setControllerAllowlist[address(_controller)] && _setControllerAllowlist[msg.sender])
```
Comment (odd-amphora) [inline contracts/JBDirectory.sol:198]: Should we use `isAllowedToSetController` in these cases?

File: test/jb_directory/add_known_controller.test.js (hunk 8 above)
Comment (odd-amphora) [reply test/jb_directory/add_known_controller.test.js:46]: Fix in all files in this PR.

File: test/jb_directory/add_known_controller.test.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    const { deployer, jbDirectory, mockJbController } = await setup();
+
+    await jbDirectory.connect(deployer).addToSetControllerAllowlist(mockJbController.address)
+
+    await expect(
+      jbDirectory.connect(deployer).addToSetControllerAllowlist(mockJbController.address)
+    ).to.revertedWith('0x30: ALREADY_ADDED');
+  });
+
+});
```
Comment (odd-amphora) [inline test/jb_directory/add_known_controller.test.js:64]: Let's add a test that adds and then removes controller(s) and ensures they are no longer allowed.

Comment (odd-amphora) [reply test/jb_directory/add_known_controller.test.js:64]: eh, i forgot about the tests below, nvm.

File: contracts/JBDirectory.sol (hunk 9 above)
Comment (drgorillamd) [reply contracts/JBDirectory.sol:198]: isAllowedToSetController is a public function -> would slightly increase gas consumption (alt being a second internal function overloading it, costing then 1U for the jump + deployment overhead, but I guess that's not really necessary here)

File: test/jb_directory/add_known_controller.test.js (hunk 8 above)
Comment (drgorillamd) [reply test/jb_directory/add_known_controller.test.js:46]: the \ is to escape the ' character, display is ok but double-checking other tests, this is indeed not consistent across them

image

File: .gitignore [hunk 11]
```diff
@@ -13,4 +13,7 @@ deployments/localhost/**
 coverage/
 coverage.json
 
-*.txt
\ No newline at end of file
+*.txt
+
+#VSCode
```
Comment (odd-amphora) [inline .gitignore:18]: super nit) space in between `#`

# VSCode

File: test/jb_directory/add_set_controller_allowed_list.test.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+
+import { deployMockContract } from '@ethereum-waffle/mock-contract';
+
+import jbController from '../../artifacts/contracts/interfaces/IJBController.sol/IJBController.json';
+import jbOperatoreStore from '../../artifacts/contracts/JBOperatorStore.sol/JBOperatorStore.json';
+import jbProjects from '../../artifacts/contracts/JBProjects.sol/JBProjects.json';
+
+describe('JBDirectory::AddToSetControllerAllowlist(...)', function () {
```
Comment (odd-amphora) [inline test/jb_directory/add_set_controller_allowed_list.test.js:10]: super nit) camel case `addToSetControllerAllowlist` (replace in other files too)

**Review by odd-amphora — APPROVED** [review]

**Review by mejango — APPROVED** [review]
Looks good, i like this pattern.

Will projects be able to add custom controllers for themselves?
